Add transition support for SmartDimmer

connorproctor opened the pull request:
* Adds a transition param to set_brightness() that specifies the duration in milliseconds that the dimmer switch will take to transition to the new brightness

This is a stab at the first part of #44.

Closes #44

kirichkov commented:
Thanks for your contribution! Can you please add some test cases so we don't break this functionality by mistake in the future?

rytilahti commented on kasa/smartdimmer.py:
```diff
@@ -41,19 +41,37 @@ def brightness(self) -> int:
     @requires_update
-    async def set_brightness(self, value: int):
-        """Set the new dimmer brightness level in percentage."""
+    async def set_brightness(self, brightness: int, transition: int = None):
```
I think it makes sense to make the transition a kwarg-only.

```suggestion
    async def set_brightness(self, brightness: int, *, transition: int = None):
```

rytilahti commented on kasa/smartdimmer.py:
```diff
@@ -41,19 +41,37 @@ def brightness(self) -> int:
+    async def set_brightness(self, brightness: int, transition: int = None):
+        """Set the new dimmer brightness level in percentage.
+
+        :param int transition: duration in milliseconds to transition to the new brightness (default: None)
```
I suppose `None` means in this case that the default transition is used? So passing `0` here would disable the transition altogether?

connorproctor replied:
Passing `0` will raise a ValueError, as the switch does not support a transition of 0.  

One subtlety between calling set_brightness() with a transition and without is that with the transition the switch automatically turns on, but without the transition the switch does not automatically turn on.

rytilahti replied:
Hmm, if the dimmers are any similar to the bulbs, i.e., there is a short transition when turning the bulb on/off, to get an immediate change (without the regular transition when nothing is defined), one should pass a very short transition?

Please add a note to the docstring that using transition will turn the device on. Having the ability to adjust the settings while the device is turned off is expected and makes sense in many cases.

rytilahti commented on kasa/smartdimmer.py:
```diff
@@ -41,19 +41,37 @@ def brightness(self) -> int:
+            raise ValueError(
+                "Brightness must be integer, " "not of %s.", type(brightness)
+            )
+        elif not 0 <= brightness <= 100:
```
Avoid `elif` construct here, as the error handling will raise an exception and it won't execute further anyway.

rytilahti commented on kasa/smartdimmer.py:
```diff
@@ -41,19 +41,37 @@ def brightness(self) -> int:
+            )
+        elif not 0 <= brightness <= 100:
+            raise ValueError("Brightness value %s is not valid." % brightness)
+        elif transition is not None and not isinstance(transition, int):
+            raise ValueError(
+                "Transition must be integer, " "not of %s.", type(transition)
+            )
+        elif transition is not None and transition <= 0:
+            raise ValueError("Transition value %s is not valid." % transition)
```
These two can be combined under `if transition is not None:` and checked under that only for the values.

rytilahti commented on kasa/smartdimmer.py:
```diff
@@ -41,19 +41,37 @@ def brightness(self) -> int:
+        if transition:
             return await self._query_helper(
-                "smartlife.iot.dimmer", "set_brightness", {"brightness": value}
+                "smartlife.iot.dimmer",
```
Move this to be a class attribute (much like done in smartbulb.py).

rytilahti commented:
@connorproctor to give you an idea about how to test that the methods are being called properly using the transition (and without), check out the linked PR.

connorproctor commented:
>btw, does the transitions also work for turn_on and turn_off somehow?

I'm not aware of a way of passing a transition duration to the switch when using `set_relay_state` via `self._query_helper("system", "set_relay_state", {"state": 1})`. Though I think it should be possible to use `set_dimmer_transition` from `turn_on` and `turn_off` with the brightness value set to the current `brightness()` or `0`, respectively.

connorproctor commented:
I'm trying to write tests but when running them on my computer all the `async` tests are skipped.

I'm getting this error:
```
async def functions are not natively supported and have been skipped.
You need to install a suitable plugin for your async framework, for example:
  - pytest-asyncio
  - pytest-trio
  - pytest-tornasync
  - pytest-twisted
```

However I do have `pytest-asyncio` installed

rytilahti commented:
That's really odd and I'm afraid I don't have any direct answer for that. Using poetry should create a working dev environment (including the necessary libs for testing, such as pytest-asyncio), but as you have it installed, I'm not really sure what is going on...

When you execute `pytest kasa/`, which plugins are you seeing in the list? Here's what I have (in a modified environment: `plugins: aiohttp-0.3.0, azurepipelines-0.8.0, requests-mock-1.7.0, anyio-1.3.1, asyncio-0.11.0, timeout-1.3.3, sugar-0.9.3, forked-1.1.3, pylint-0.14.1, cov-2.9.0, mock-3.1.1, xdist-1.32.0`). I think the `asyncio-0.11.0` is for pytest-asyncio here.

connorproctor commented:
With a fresh poetry environment these were my pytest plugins when running `poetry run pytest kasa/`:
`plugins: azurepipelines-0.8.0, cov-2.9.0, mock-3.1.1, sugar-0.9.3, anyio-1.3.1, asyncio-0.11.0`.

I manually installed all the plugins you have listed and now most of the tests run. Thanks for the help. This is my first time using poetry, so I'm not sure where the problem is.

rytilahti commented on kasa/smartdimmer.py:
```diff
@@ -41,19 +43,42 @@ def brightness(self) -> int:
+        """Set the new dimmer brightness level in percentage.
+
+        :param int transition: transition duration in milliseconds.
+            Will turn on the light if brightness > 0,
```
I think this should explicitly state that these will happen only if `transition` is used.

rytilahti commented on kasa/tests/test_dimmer.py:
```diff
@@ -0,0 +1,36 @@
+        "smartlife.iot.dimmer",
+        "set_dimmer_transition",
+        {"brightness": 99, "duration": 1000},
+    )
```
It would be great to have also tests for the implicit turn on / turn off when using the transition.

rytilahti commented:
Doing `poetry install` should create automatically a working environment (that can be accessed using `poetry run`), so there should be no need for manual installations.. Glad you got it working though!

Anyway, in case you had already activated virtualenv, the install will also use that (that's how I have my own dev setup).

connorproctor commented:
I'm not sure if it's something weird with my local setup or not, but it looks like the missing dependency that fix the tests was `pytest-aiohttp`.  To verify I deleted the virtual environment, did a fresh `poetry install`, ran `poetry run pytest` to verify tests were broken, ran `poetry run pip install pytest-aiohttp`, reran `poetry run pytest` to verify that most tests were now working. 

I updated the tests, let me know what you think.

rytilahti commented on kasa/tests/test_dimmer.py:
```diff
@@ -0,0 +1,60 @@
+
+    await dev.set_brightness(0, transition=1000)
+
+    assert dev.brightness == original_brightness
```
This cannot be correct, right? I think it simply requires `await dev.update()` to update the cache, otherwise it's using the old (pre set_brightness) state.

connorproctor replied:
This is correct, when setting to brightness of 0 via a transition it resets the brightness to the original brightness after turning off.

connorproctor replied:
I'm not sure how people use this library, so I'm not sure the right API abstraction. Maybe a more logical thing to do would be to disallow a brightness of `0` in `set_brightness()` and to instead add a transition to `turn_off()`.

rytilahti replied:
Ah, I see. If you don't mind, add a short note about the behavior.

About the `turn_off` vs `set_brightness(0)`, in my mind it is better to be explicit when the device is turned off. For the bulbs, 0 is also a valid brightness level while still being on, so in my opinion the dimmer should behave the same (setting 0 should set it to the lowest brightness level), `turn_off` should be used to turn the device off.

connorproctor replied:
I just tried calling `set_brightness(0)` and it doesn't actually work. Looks like PR #47 was a misguided (or the dimmer switch behaves differently from the bulbs)

```
DEBUG:kasa.protocol:> (63) {"smartlife.iot.dimmer": {"set_brightness": {"brightness": 0}}}
DEBUG:kasa.protocol:< (88) {'smartlife.iot.dimmer': {'set_brightness': {'err_code': -3,
                                             'err_msg': 'invalid argument'}}}
```

connorproctor replied:
I pushed another commit that has a different take on how to approach this: https://github.com/python-kasa/python-kasa/commit/9f336ee005f747c36291e34be6d99665041bfced

- Add `transition` support to turn_on and turn_off 
- Disallow brightness 0 in `set_brightness()`, since it doesn't work if there's not a transition and is confusing if there is.

rytilahti replied:
Having the `transition` for `turn_{on,off}` sounds good (you need to add ignore for the type error to make mypy pass, see my bulb transition PR).

I just tested my KL60 and for that 0 is the lowest brightness level it has, so I'm assuming the rest of the bulbs do the same. Therefore I would suggest simply coercing the 0 brightness to 1 for dimmers, then the interface would function the same for both bulbs and dimmers (for homeassistant, these are both handled as lights, so having a common interface is preferable).

rytilahti commented:
> I'm not sure if it's something weird with my local setup or not, but it looks like the missing dependency that fixed the tests was `pytest-aiohttp`. To verify I: deleted the virtual environment, did a fresh `poetry install`, ran `poetry run pytest` to verify tests were broken, ran `poetry run pip install pytest-aiohttp`, reran `poetry run pytest` to verify that most tests were now working.

Indeed, I did also check that out and looks like it is not working, and the same thing also happens on CI for other python versions than what I have locally (3.8), but as it's a warning I never noticed that...

This is really odd, I'm curious what causes it to work when installing pytest-aiohttp (which I confirmed is the case)? As we are not using aiohttp in this library, adding its pytest plugin as a dependency feels wrong. A short googling reveals this https://github.com/pytest-dev/pytest/pull/7144 but we are not really using unittest here, either..
